File: backend/app/schemas/sources.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class DataSourceRead(BaseModel):
    id: str
    workspace_code: str
    domain_code: str
    source_type: str
    name: str
    url: str | None
    enabled: bool
    default_focus_id: int
    backfill_days: int
    source_score: float
    last_fetch_at: datetime | None
    last_success_at: datetime | None
    last_error: str
    primary_category: str
    info_category: str
    source_tags: list[str] = Field(default_factory=list)
    source_secondary_tags: list[str] = Field(default_factory=list)
    source_tier: str = ""
    source_channel_type: str = ""
    expert_routes: list[str] = Field(default_factory=list)
    inclusion_recommendation: str = ""
    metadata_only: bool = False
    needs_entry: bool = False
    fetch_entry_status: str = ""
    source_quality_notes: str = ""
    workspace_link_enabled: bool | None = None
    workspace_source_weight: float | None = None
    workspace_daily_limit: int | None = None
    workspace_clustering_config: dict[str, Any] = Field(default_factory=dict)


class DataSourceRecentRawRead(BaseModel):
    id: str
    source_title: str
    source_url: str | None
    raw_content_excerpt: str
    fetched_at: datetime
    published_at: datetime | None


class DataSourceRunSummaryRead(BaseModel):
    run_id: str
    run_key: str
    run_type: str
    status: str
    completed_at: datetime | None
    fetched: int
    created: int
    updated: int
    error: str


class DataSourceTrendPointRead(BaseModel):
    day_key: str
    raw_count: int


class DataSourceDetailRead(BaseModel):
    source: DataSourceRead
    raw_count: int
    news_count: int
    recent_raw_items: list[DataSourceRecentRawRead] = Field(default_factory=list)
    recent_runs: list[DataSourceRunSummaryRead] = Field(default_factory=list)
    error_logs: list[DataSourceRunSummaryRead] = Field(default_factory=list)
    raw_trend: list[DataSourceTrendPointRead] = Field(default_factory=list)


class DataSourceWorkspaceConfigUpdate(BaseModel):
    workspace_code: str
    enabled: bool
    source_weight: float = Field(default=1.0, ge=0)
    daily_limit: int | None = Field(default=None, ge=0)


CUSTOM_SOURCE_TYPES = ["rss", "paper_rss", "paper_api", "page_manual", "page_monitor"]


class DataSourceCreate(BaseModel):
    workspace_code: str
    name: str = Field(min_length=1, max_length=255)
    source_type: str = "rss"
    url: str = Field(min_length=1)
    domain_code: str = Field(default="ai", min_length=1, max_length=64)
    backfill_days: int = Field(default=7, ge=0)
    source_weight: float = Field(default=1.0, ge=0)
    daily_limit: int | None = Field(default=None, ge=0)
    reuse_existing: bool = True


class DataSourceCreateRead(BaseModel):
    source: DataSourceRead
    created: bool


class DataSourceDefinitionUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    url: str | None = None
    enabled: bool | None = None
    backfill_days: int | None = Field(default=None, ge=0)


class LegacySeedImportRead(BaseModel):
    created: int
    updated: int
    total: int


class TechInsightLoopImportRead(BaseModel):
    created: int
    updated: int
    total: int
    fetchable: int
    metadata_only: int


class SourceImportPreviewSampleRead(BaseModel):
    name: str
    source_type: str
    url: str | None


class SourceImportPreviewRead(BaseModel):
    catalog: str
    total: int
    would_create: int
    would_update: int
    samples: list[SourceImportPreviewSampleRead] = Field(default_factory=list)


class SourceFetchRead(BaseModel):
    data_source_id: str
    source_type: str
    fetched: int
    created: int
    updated: int

File: frontend/src/api/sources.ts
import { requestJson } from "./http";

export interface DataSourceRecord {
  id: string;
  workspace_code: string;
  domain_code: string;
  source_type: string;
  name: string;
  url: string | null;
  enabled: boolean;
  default_focus_id: number;
  backfill_days: number;
  source_score: number;
  last_fetch_at: string | null;
  last_success_at: string | null;
  last_error: string;
  primary_category: string;
  info_category: string;
  source_tags: string[];
  source_secondary_tags: string[];
  source_tier: string;
  source_channel_type: string;
  expert_routes: string[];
  inclusion_recommendation: string;
  metadata_only: boolean;
  needs_entry: boolean;
  fetch_entry_status: string;
  source_quality_notes: string;
  workspace_link_enabled: boolean | null;
  workspace_source_weight: number | null;
  workspace_daily_limit: number | null;
  workspace_clustering_config: Record<string, unknown>;
}

export interface SourceRecentRawRecord {
  id: string;
  source_title: string;
  source_url: string | null;
  raw_content_excerpt: string;
  fetched_at: string;
  published_at: string | null;
}

export interface SourceRunSummaryRecord {
  run_id: string;
  run_key: string;
  run_type: string;
  status: string;
  completed_at: string | null;
  fetched: number;
  created: number;
  updated: number;
  error: string;
}

export interface SourceTrendPointRecord {
  day_key: string;
  raw_count: number;
}

export interface SourceDetailRecord {
  source: DataSourceRecord;
  raw_count: number;
  news_count: number;
  recent_raw_items: SourceRecentRawRecord[];
  recent_runs: SourceRunSummaryRecord[];
  error_logs: SourceRunSummaryRecord[];
  raw_trend: SourceTrendPointRecord[];
}

export interface LegacySeedImportResult {
  created: number;
  updated: number;
  total: number;
}

export interface TechInsightLoopImportResult {
  created: number;
  updated: number;
  total: number;
  fetchable: number;
  metadata_only: number;
}

export interface SourceImportPreviewSample {
  name: string;
  source_type: string;
  url: string | null;
}

export interface SourceImportPreview {
  catalog: string;
  total: number;
  would_create: number;
  would_update: number;
  samples: SourceImportPreviewSample[];
}

export interface SourceWorkspaceConfigUpdate {
  workspace_code: string;
  enabled: boolean;
  source_weight: number;
  daily_limit: number | null;
}

export interface SourceFetchResult {
  data_source_id: string;
  source_type: string;
  fetched: number;
  created: number;
  updated: number;
}

export interface SourceCreatePayload {
  workspace_code: string;
  name: string;
  source_type: string;
  url: string;
  domain_code?: string;
  backfill_days?: number;
  source_weight?: number;
  daily_limit?: number | null;
  reuse_existing?: boolean;
}

export interface SourceCreateResult {
  source: DataSourceRecord;
  created: boolean;
}

export interface SourceDefinitionUpdatePayload {
  name?: string;
  url?: string;
  enabled?: boolean;
  backfill_days?: number;
}

export async function fetchSources(workspaceCode?: string): Promise<DataSourceRecord[]> {
  const params = workspaceCode ? `?${new URLSearchParams({ workspace_code: workspaceCode }).toString()}` : "";
  return requestJson<DataSourceRecord[]>(`/api/sources${params}`);
}

export async function fetchSourceDetail(sourceId: string, workspaceCode?: string): Promise<SourceDetailRecord> {
  const params = workspaceCode ? `?${new URLSearchParams({ workspace_code: workspaceCode }).toString()}` : "";
  return requestJson<SourceDetailRecord>(`/api/sources/${sourceId}${params}`);
}

export async function previewSourceImport(catalog: "legacy" | "tech"): Promise<SourceImportPreview> {
  const params = new URLSearchParams({ catalog });
  return requestJson<SourceImportPreview>(`/api/sources/import-preview?${params.toString()}`);
}

export async function importLegacySources(): Promise<LegacySeedImportResult> {
  return requestJson<LegacySeedImportResult>("/api/sources/import-legacy-seeds", {
    method: "POST"
  });
}

export async function importTechInsightLoopSources(): Promise<TechInsightLoopImportResult> {
  return requestJson<TechInsightLoopImportResult>("/api/sources/import-tech-insight-loop", {
    method: "POST"
  });
}

export async function fetchSource(sourceId: string, workspaceCode?: string): Promise<SourceFetchResult> {
  const params = workspaceCode ? `?${new URLSearchParams({ workspace_code: workspaceCode }).toString()}` : "";
  return requestJson<SourceFetchResult>(`/api/sources/${sourceId}/fetch${params}`, {
    method: "POST"
  });
}

export async function createSource(payload: SourceCreatePayload): Promise<SourceCreateResult> {
  return requestJson<SourceCreateResult>("/api/sources", {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export async function updateSourceDefinition(
  sourceId: string,
  payload: SourceDefinitionUpdatePayload,
  workspaceCode?: string
): Promise<DataSourceRecord> {
  const params = workspaceCode ? `?${new URLSearchParams({ workspace_code: workspaceCode }).toString()}` : "";
  return requestJson<DataSourceRecord>(`/api/sources/${sourceId}${params}`, {
    method: "PATCH",
    body: JSON.stringify(payload)
  });
}

export async function updateSourceWorkspaceConfig(
  sourceId: string,
  payload: SourceWorkspaceConfigUpdate
): Promise<DataSourceRecord> {
  return requestJson<DataSourceRecord>(`/api/sources/${sourceId}/workspace-link`, {
    method: "PATCH",
    body: JSON.stringify(payload)
  });
}
